tests/unit/s2n_client_psk_extension_test.c

```diff
@@ -1543,5 +1545,144 @@ int main(int argc, char **argv)
+        /* Set the ticket issue time to be 10 milliseconds ago, so the obfuscated ticket age will be non-zero */
+        struct s2n_psk *psk = client_conn->psk_params.chosen_psk;
+        psk->ticket_issue_time -= MILLIS_TO_NANOS(10);
+
+        /* Calculate the size of the identity without the key name. Used to first seek to the key name as the target for
+         * s2n_stuffer_skip_read_until, and then seek to the end of the identity. */
+        const unsigned long identity_data_size = psk->identity.size - strlen((char*) psk->identity.data);
+
+        /* ClientHello 1 */
+        EXPECT_SUCCESS(s2n_client_hello_send(client_conn));
+
+        /* Read the obfuscated ticket age from ClientHello 1 */
+        EXPECT_SUCCESS(s2n_stuffer_skip_read_until(client_io, (char*) psk->identity.data));
+        EXPECT_SUCCESS(s2n_stuffer_skip_read(client_io, identity_data_size));
+        uint32_t obfuscated_ticket_age_1 = 0;
+        EXPECT_SUCCESS(s2n_stuffer_read_uint32(client_io, &obfuscated_ticket_age_1));
+
+        /* Skip over the size of the binders list */
+        EXPECT_SUCCESS(s2n_stuffer_skip_read(client_io, sizeof(uint16_t)));
+
+        /* Ensure the binder size is as expected */
+        uint8_t binder_1_size = 0;
+        EXPECT_SUCCESS(s2n_stuffer_read_uint8(client_io, &binder_1_size));
+        EXPECT_TRUE(binder_1_size == BINDER_SIZE);
+
+        /* Read the binder from ClientHello 1 */
+        uint8_t binder_1[BINDER_SIZE] = { 0 };
+        EXPECT_SUCCESS(s2n_stuffer_read_bytes(client_io, binder_1, BINDER_SIZE));
+        EXPECT_SUCCESS(s2n_stuffer_reread(client_io));
+
+        /* Ensure that the ticket age and binder are non-zero */
+        EXPECT_TRUE(obfuscated_ticket_age_1 != 0);
+        uint8_t zero_array[BINDER_SIZE] = { 0 };
+        EXPECT_FALSE(s2n_constant_time_equals(binder_1, zero_array, BINDER_SIZE));
+
+        EXPECT_SUCCESS(s2n_stuffer_copy(client_io, server_io,
+                                        s2n_stuffer_data_available(client_io)));
+
+        /* Server receives ClientHello 1 */
+        EXPECT_SUCCESS(s2n_client_hello_recv(server_conn));
+        EXPECT_SUCCESS(s2n_set_connection_hello_retry_flags(server_conn));
+
+        /* Server sends HelloRetryRequest */
+        EXPECT_SUCCESS(s2n_server_hello_retry_send(server_conn));
+
+        EXPECT_SUCCESS(s2n_stuffer_wipe(client_io));
+        EXPECT_SUCCESS(s2n_stuffer_copy(server_io, client_io,
+                                        s2n_stuffer_data_available(server_io)));
+        client_conn->handshake.message_number = 1; /* hello retry */
+
+        /* Client receives HelloRetryRequest */
+        EXPECT_SUCCESS(s2n_server_hello_recv(client_conn));
+
+        EXPECT_SUCCESS(s2n_stuffer_rewrite(client_io));
+
+        /* Change the ticket issue time so a new obfuscated ticket age will change */
+        psk->ticket_issue_time -= MILLIS_TO_NANOS(1);
+
+        /* Client sends ClientHello 2 */
+        EXPECT_SUCCESS(s2n_client_hello_send(client_conn));
```

Comment:
Could all of this be replaced by just continuing the handshake with s2n_negotiate_test_server_and_client?

Reply:
I'm not sure I can use the negotiate functions here since it seems like the stuffers are wiped at the end of `s2n_handshake_io.c`, so I don't think I'll be able to search through the stuffers:

https://github.com/aws/s2n-tls/blob/1349ead0dba975b5bf191310d47b4c6d3aae85a5/tls/s2n_handshake_io.c#L1007-L1008

Reply:
I changed this test to use `s2n_connection_set_io_stuffers` in 7670344c3cead402e78809c88ada79d6476b9993 so the negotiate calls can be used and still retain the data written to the io stuffers.